### ISSUE TITLE
Future: migrate --device insane from HF pipeline to WhisperX

### ISSUE THREAD

**Issue by zackees** [zackees/transcribe-anything#81]
## Background

We empirically confirmed in [this triage session](https://github.com/zackees/transcribe-anything/issues/40) that the timing delta between \`--device insane\` and \`--device cuda\` is **algorithmic, not sample-rate-driven**:

- \`--device cuda\` (openai-whisper) uses **sequential** 30-s decoding with cumulative seek offsets.
- \`--device insane\` (insanely-fast-whisper → HuggingFace \`pipeline(\"automatic-speech-recognition\")\`) uses **chunked-parallel** decoding with stride-based stitching, which inherits a long-standing tokenizer offset bug (HF transformers [#34472](https://github.com/huggingface/transformers/issues/34472), [#34210](https://github.com/huggingface/transformers/issues/34210), [#31942](https://github.com/huggingface/transformers/issues/31942)).

The transformers pin bump that landed alongside this issue (\`transformers==4.55.4\`) picks up the upstream fixes for the immediate drift on >30 s audio, but Whisper's native segment-level timestamps are still inherently coarse (drift by hundreds of ms across long audio) regardless of which decoder is used.

## Proposal

Replace the HF-pipeline backend behind \`--device insane\` with **[WhisperX](https://github.com/m-bain/whisperX)**, which:

1. Uses **faster-whisper** (CTranslate2) under the hood for transcription — fast like the current path.
2. Adds **wav2vec2 forced alignment** on top, producing **phoneme-precise word-level timestamps** that are immune to chunk drift entirely.
3. Has built-in **VAD-based chunk boundaries** so chunks fall on silences instead of arbitrary 30-s lines.
4. Already integrates pyannote.audio diarization, which we currently invoke separately.

## Scope

- New dependency tree (no longer needs \`insanely-fast-whisper==0.0.15\`, \`transformers\`, or the HF pipeline path).
- Same input/output contract: pre-resampled 16 kHz WAV in, \`out.{srt,vtt,txt,json}\` out.
- Diarization migrates from our current bespoke path (HF token + speechbrain) to WhisperX's built-in pyannote integration.
- Test: existing \`test_insane_vs_cuda_longform_timestamps.py\` should pass with even tighter tolerance (well under 1 s on long-form).

## Trade-offs

| Concern             | Current (HF pipeline)                          | WhisperX                                                 |
|---------------------|-------------------------------------------------|----------------------------------------------------------|
| Timestamp accuracy  | segment-level, drifts on long audio             | word-level, phoneme-aligned, drift-free                  |
| Word-level support  | flaky via \`return_timestamps='word'\`            | first-class                                              |
| Diarization         | separate speechbrain+k2 path (needed #69 stub) | bundled                                                  |
| Install footprint   | ~3 GB (transformers + insanely-fast-whisper)    | ~3 GB (faster-whisper + wav2vec2 alignment models)       |
| First-run download  | model + transformers wheels                     | model + wav2vec2 + diarization (similar)                 |

## Out of scope here

- Decision deferred — needs validation that WhisperX's alignment models work cleanly on the same Linux+Windows+macOS matrix we support today.

Filed by Claude Code during triage of the timestamp drift question. Not blocking any current user issue, but the cleanest long-term solution to the entire \`--device insane\` timing-accuracy story.

**Comment by zackees** [comment]
Triage note — re-reading this with current `main` state in mind:

Since this was filed (2026-05-30), `--device whisperx` shipped as a **parallel, additive** backend (PR #87, commit 666c30d) and `--device insane-flash` shipped right after (PR #90). So the original framing of \"migrate `--device insane` from HF pipeline to WhisperX\" is now ambiguous between two very different proposals:

1. **\"Make `--device insane` an alias for the WhisperX backend\"** — i.e., silently swap the implementation. This is a **breaking change**, not a refactor: `insanely_fast_whisper.py` emits the HF pipeline shape `{\"chunks\":[{\"timestamp\":[s,e],\"text\":...}], \"text\":...}`, while `whisperx.py` emits `{\"segments\":[{\"start\",\"end\",\"text\",\"words\":[...],\"speaker\":...}], ...}`. The auxiliary outputs (`out.srt/vtt/txt`) line up, but any downstream consumer reading `out.json` directly would break.

2. **\"Deprecate `--device insane` and recommend `--device whisperx`\"** — a policy/EOL question. The recent investment in fixing the insane backend (#74 cusparselt, #76 k2 stub, #79 banner, #83 transformers bump) and the deliberate \"additive\" language in PR #87 suggest the maintainer is keeping both. But there's no recorded policy.

Tagging `triage:more-discussion`. Concrete questions to resolve before this can move:

- Is `--device insane` maintenance-only with `--device whisperx` as the recommended path, or are both first-class indefinitely?
- If maintenance-only, what's the EOL criterion (transformers/insanely-fast-whisper going stale, a specific release, a date)?
- If a silent swap is on the table, what's the deprecation period and `out.json` migration story?

If the answer is \"keep both as first-class, no migration planned,\" the cleanest resolution is to close this issue as superseded by #87/#90.

cc @zackees for the policy call.

**Comment by zackees** [comment]
## Investigation: alignment-only path on top of `--device insane` / `--device insane-flash`

Picking up the open question from my prior comment — **\"can we keep `--device insane` (and `--device insane-flash`) and just borrow WhisperX's timing accuracy?\"** Answer: **yes, cleanly, with zero new dependencies.**

### What WhisperX's timing improvement actually is

WhisperX's accuracy advantage on timestamps doesn't come from its transcription engine (`faster-whisper` via CTranslate2) — it comes from a **separate wav2vec2 forced-alignment pass** that runs *after* the transcript is generated. Given `(audio, [{start, end, text}, ...])`, it aligns the recognized text against the audio waveform at the phoneme level and produces per-word `{word, start, end, score}` timestamps that are immune to Whisper's chunk-stride drift.

That alignment step is implemented as two public Python functions on the `whisperx` package:

```python
# whisperx/alignment.py @ 3.8.6
def load_align_model(language_code: str, device: str, model_name=None, ...) -> (torch.nn.Module, dict): ...
def align(transcript, model, align_model_metadata, audio, device, ...) -> AlignedTranscriptionResult: ...
```

Neither touches a transcription model. They run against arbitrary `{start, end, text}` segments. Confirmed against the current `main` branch source ([whisperX/whisperx/alignment.py](https://github.com/m-bain/whisperX/blob/main/whisperx/alignment.py)), and there's external precedent ([`jim60105/docker-whisperX/load_align_model.py`](https://github.com/jim60105/docker-whisperX/blob/master/load_align_model.py), whisperx issues #939, #1308) for using these in isolation.

### Proposed UX

Opt-in flag, no behavior change without it:

```bash
transcribe-anything input.mp3 --device insane --align
transcribe-anything input.mp3 --device insane-flash --align
```

With `--align`:
- `out.json` grows a `words: [{word, start, end, score}]` array per segment and tightens segment-level `start`/`end` to the first/last aligned word.
- `out.srt` / `out.vtt` get tightened segment timestamps (the original chunk drift goes away). Word-level karaoke highlighting is a follow-up.
- Backwards-compat: without `--align`, output is bit-for-bit identical to today.

### Implementation sketch

1. Reuse the already-pinned **`whisperx==3.8.6` iso-env** from `src/transcribe_anything/whisperx_reqs.py`. No new dependency stack in the insane env.
2. New module `src/transcribe_anything/insane_align_runner.py` — runs inside the whisperx iso-env. Reads insane's `out.json`, converts `chunks` → `segments`, calls `load_align_model` + `align`, writes back enriched JSON.
3. New module `src/transcribe_anything/insane_align.py` — host-side wrapper. Loads the whisperx iso-env, invokes the runner.
4. `_cmd.py`: add `--align` flag, gated to `--device insane` and `--device insane-flash`.
5. `insanely_fast_whisper.run_insanely_fast_whisper()`: post-process step at the end of the function when `align=True`.

Input/output shape adaptation:

| insane JSON (current)                                  | WhisperX `align()` input                           |
|--------------------------------------------------------|----------------------------------------------------|
| `{\"chunks\":[{\"timestamp\":[s_start, s_end], \"text\":...}]}` | `[{\"start\": float_s, \"end\": float_s, \"text\": str}]`     |

One comprehension, plus the existing sanitization for `None` end timestamps (`trim_text_chunks` + the `_last_end_timestamp_from_insane_json` test helper already encode this pattern).

### Edge cases to handle

1. **Unsupported language.** WhisperX ships defaults for 41 languages (5 torchaudio + 36 HF wav2vec2). Unknown language code → `ValueError`. Guard with the `DEFAULT_ALIGN_MODELS_*` dict and emit a warning + skip alignment (output unchanged) rather than fail the whole transcription.
2. **Missing language.** When the user didn't pass `--language`, insane's pipeline emits a top-level `language` field — forward it to `load_align_model`.
3. **`None` end-timestamps.** Already-known pattern in the insane chunks; reuse the existing fallback.
4. **NLTK `punkt_tab` download.** `align()` lazy-downloads NLTK punkt on first run. Either pre-warm in `cli_init_insane*` or document.
5. **GPU contention.** Alignment runs on the same `device` as insane. It's a separate forward pass after transcription finishes, so VRAM isn't contested. Empirically the alignment pass is 5–15x faster than transcription on the same GPU.

### Cost estimate

~250 LOC, 2 new files, ~10 new unit tests for the JSON shape conversion (full end-to-end needs CUDA + the whisperx iso-env built once).

### Recommendation

This unblocks the original #81 motivation (\"insane's segment timestamps drift on long-form audio\") **without** deprecating `--device insane`/`--device insane-flash`, and **without** adding a new heavyweight dep tree. The whisperx env is already there for `--device whisperx` users; this just reuses it.

If you're good with the `--align` flag plan, I can stage it as a follow-up PR. Otherwise the alternative answers to my prior comment still stand:

- *Keep both as first-class, no migration planned* → close #81 as superseded by #87 / #90.
- *Deprecate `--device insane`* → still on the table but no longer necessary if `--align` lands.

cc @zackees for the call.

---

**Sources reviewed during this investigation:**
- [whisperX/whisperx/alignment.py](https://github.com/m-bain/whisperX/blob/main/whisperx/alignment.py) — `load_align_model` and `align` signatures, default-models dict, `ValueError` on unsupported language
- [whisperX/whisperx/schema.py](https://github.com/m-bain/whisperX/blob/main/whisperx/schema.py) — `SingleSegment` and `AlignedTranscriptionResult` shapes
- [whisperX README — Python usage](https://github.com/m-bain/whisperX/blob/main/README.md) — confirms `align()` can be fed segments from non-whisperx sources
- [jim60105/docker-whisperX/load_align_model.py](https://github.com/jim60105/docker-whisperX/blob/master/load_align_model.py) — standalone reference usage
- This repo's `src/transcribe_anything/insanely_fast_whisper.py` (chunks/timestamp JSON shape), `src/transcribe_anything/whisperx_reqs.py` (the env we'd reuse), and `tests/test_insane_vs_cuda_longform_timestamps.py` (the regression we'd want to tighten)

**Comment by zackees** [comment]
Closing — the underlying motivation (timing accuracy on `--device insane` / `--device insane-flash`) is now addressed by **PR #102, merged**, which adds an opt-in `--align` post-pass that runs WhisperX's wav2vec2 forced alignment on the insane transcript. No new deps in the insane env, no behavior change without the flag, and graceful fallback when the language is outside WhisperX's 41 defaults.

What you can do now:

```bash
transcribe-anything input.mp3 --device insane --align
transcribe-anything input.mp3 --device insane-flash --align
transcribe-anything input.mp3 --device insane --align --align_model facebook/wav2vec2-large-960h-lv60-self
```

`out.json` grows `chunks[i].words = [{word, start, end, score}]` and tightens `chunks[i].timestamp`; `out.srt` / `out.vtt` inherit the tightened segment bounds.

The original \"migrate insane to whisperx\" framing is no longer needed — both backends stay first-class. If a future case emerges for actually retiring `--device insane` (e.g., upstream `insanely-fast-whisper` going stale, or a user demand we can't otherwise meet), that's a separate conversation under a new issue.

Reference: [investigation comment](https://github.com/zackees/transcribe-anything/issues/81#issuecomment-4617999777), [PR #102](https://github.com/zackees/transcribe-anything/pull/102).
